core/src/main/java/org/apache/iceberg/rest/ScanTaskIterable.java

```diff
@@ -143,10 +141,14 @@ public void run() {
         int remaining = activeWorkers.decrementAndGet();
 
         if (remaining == 0
-            && !planTasks.isEmpty()
+            && planTasks.isEmpty()
             && !shutdown.get()
-            && !initialFileScanTasks.isEmpty()) {
-          executorService.execute(new PlanTaskWorker());
+            && initialFileScanTasks.isEmpty()) {
+          try {
+            taskQueue.put(DUMMY_TASK);
+          } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+          }
```

Comment:
Here is one race i can think on this context, what if the worker increment and get returned 0 and some one called cancel which will mark (shutdown get as true) and in that scenario we would skip adding the poison pill to the task and the iterator hasNext() never return false it might be stuck, but i think we can argue who calls hasNext on a closed iterator ?